### ISSUE TITLE
ActiveRecord::ConnectionNotEstablished For rails multi DB

### ISSUE THREAD

**Issue by uxxman** [bensheldon/good_job#1103]
First of all great work 💟 

I have a basic multi db setup with good_job added to the primary DB. When starting the server and making first request, GoodJob is not able to subscribe with LISTEN with error stating  `GoodJob error: ActiveRecord::ConnectionNotEstablished: No connection pool for 'ActiveRecord::Base' found.`

Also tried the active_record_parent_class but error persist
`GoodJob.active_record_parent_class = 'ApplicationRecord'`

The issue looks like that good_job is trying to connect too early, even rails hasn't checked schema migrations yet since scheme check queries are logged after this error.

### Repo to reproduce
https://github.com/uxxman/good_job_multi_db
Its the most basic multi db setup from rails guide and then good_job added to primary db

### Note:
The error delays first server request for 20+ seconds, and after that good_job is able to find and connect to DB, since rails has set it up till this point and from there onwards, all works as expected.

**Comment by bensheldon** [comment]
@uxxman thank you for the reproduction! I'll take a look 👀

**Comment by bensheldon** [comment]
I'm not super familiar with how multiple databases work, but this change seems to fix it:

```diff
# app/models/application_record.rb
  class ApplicationRecord < ActiveRecord::Base
    primary_abstract_class
+   connects_to database: { writing: :primary, reading: :primary }
  end
```

My assumption is that when using multiple databases, you must be explicit of which database to use in the Base/Application Record

**Comment by uxxman** [comment]
@bensheldon thank you for looking into it. I have added explicit connects_to to the repo https://github.com/uxxman/good_job_multi_db/commit/0b1f3f5084203b21ff3c0de8d4ca132ef51fdb1b

But the issue still persists. 

My understanding is that good_jobs hooks into activerecord too early even before rails does, which is cause this issue on just initial boot up and first request.....later everything works as expected.

**Comment by bensheldon** [comment]
Nuts, sorry about that. I now realize I had something else commented out so I shouldn't have been so confident that was the problem.

I'm still not quite sure what is going on specifically with multiple databases. I noticed this is also Rails 7.1, so that's new!

Here's the initialization code:

https://github.com/bensheldon/good_job/blob/7c5efa074db7f87d9296582ad54cab9ef2cba196/lib/good_job/engine.rb#L51-L69

**Comment by bensheldon** [comment]
Two observations:

1. This does seem to work (unless I'm overlooking something) if you create an additional `class PrimaryRecord < ApplicationRecord` and then set `GoodJob.active_record_parent_class = 'PrimaryRecord'`. I'm not sure if that implies that I don't understand going on with the Active Record class hierarchy with multiple databases.
2. Alternatively, I think the much better way to configure GoodJob is to directly extend the class, rather than try to change its parent class:

    ```ruby
    # config/initializers/good_job.rb
    ActiveSupport.on_load(:good_job_base_record) do
      connects_to database: { writing: :primary, reading: :primary }
    end
    ```

**Comment by uxxman** [comment]
Looking at engine file and intializer code. we can make use of the server block which achieves what we need to in a better way. This solves the issue as well.....I have a setup a PR for this can you have a look https://github.com/bensheldon/good_job/pull/1106

**Comment by eileencodes** [comment]
This should be fixed with https://github.com/rails/rails/pull/49701 which was backported to 7-1-stable. It will be included in the next release of 7.1.
